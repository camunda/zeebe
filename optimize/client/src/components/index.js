/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Analysis} from './Analysis';
export {Footer} from './Footer';
export {Header} from './Header';
export {PrivateRoute} from './PrivateRoute';
export {Dashboard} from './Dashboards';
export {Report} from './Reports';
export {Sharing} from './Sharing';
export {Home, Collection} from './Home';
export {default as Logout} from './Logout';
export {Processes, ProcessReport} from './Processes';
