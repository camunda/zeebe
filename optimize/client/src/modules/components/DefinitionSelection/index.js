/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {default as DefinitionSelection} from './DefinitionSelection';
export {default as VersionPopover} from './VersionPopover';
export {default as TenantPopover} from './TenantPopover';
