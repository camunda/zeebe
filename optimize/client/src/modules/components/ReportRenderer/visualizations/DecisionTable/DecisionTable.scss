@import 'shared-styles/shared-styles.scss';

.DecisionTable {
  @include scrollable;

  .dmn-decision-table-container .tjs-container {
    width: 100%;

    table {
      width: 100%;

      thead {
        .annotation {
          width: 192px !important;
        }
      }
    }
  }

  tfoot td:first-child {
    padding-left: 5px;
    border-right: solid 1px var(--table-cell-border-color);
  }
}
