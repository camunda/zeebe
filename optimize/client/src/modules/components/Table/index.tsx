/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {default as Table} from './Table';

export type {Body as TableBody, Head as TableHead} from './Table';
