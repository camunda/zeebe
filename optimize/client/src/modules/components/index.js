/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Table} from './Table';
export {Button} from './Button';
export {Modal} from './Modal';
export {Dropdown} from './Dropdown';
export {Deleter} from './Deleter';
export {Input} from './Input';
export {Select} from './Select';
export {BPMNDiagram, ClickBehavior, PartHighlight} from './BPMNDiagram';
export {DMNDiagram} from './DMNDiagram';
export {CopyToClipboard} from './CopyToClipboard';
export {ReportRenderer, NoDataNotice} from './ReportRenderer';
export {ErrorBoundary} from './ErrorBoundary';
export {Icon} from './Icon';
export {ActionItem} from './ActionItem';
export {ShareEntity} from './ShareEntity';
export {DashboardRenderer, DiagramScrollLock} from './DashboardRenderer';
export {DefinitionSelection, VersionPopover, TenantPopover} from './DefinitionSelection';
export {TargetValueBadge} from './TargetValueBadge';
export {ErrorPage} from './ErrorPage';
export {Labeled} from './Labeled';
export {LabeledInput} from './LabeledInput';
export {DatePicker} from './DatePicker';
export {LoadingIndicator} from './LoadingIndicator';
export {ColorPicker} from './ColorPicker';
export {EntityName} from './EntityName';
export {EntityNameForm} from './EntityNameForm';
export {SubNav, NavItem} from './navigation';
export {HeatmapOverlay} from './HeatmapOverlay';
export {EntityList} from './EntityList';
export {UserTypeahead} from './UserTypeahead';
export {DateRangeInput} from './DateRangeInput';
export {LastModifiedInfo} from './LastModifiedInfo';
export {ReportDetails} from './ReportDetails';
export {InstanceCount} from './InstanceCount';
export {DocsLink} from './DocsLink';
export {Tabs} from './Tabs';
export {Checklist} from './Checklist';
export {Tooltip} from './Tooltip';
export {SelectionPreview} from './SelectionPreview';
export {DownloadButton} from './DownloadButton';
export {BulkDeleter} from './BulkDeleter';
export {AlertModal} from './AlertModal';
export {AlertsDropdown} from './AlertsDropdown';
export {DurationChart} from './DurationChart';
export {ReportTemplateModal, DashboardTemplateModal} from './TemplateModal';
export {KpiCreationModal} from './KpiCreationModal';
export {EmptyState} from './EmptyState';
export {PageTitle} from './PageTitle';
export {TextEditor} from './TextEditor';
export {EntityDescription} from './EntityDescription';
export {Popover} from './Popover';
export {InstanceViewTable} from './InstanceViewTable';
export {TenantInfo} from './TenantInfo';
export {TargetSelection} from './TargetSelection';
export {Loading} from './Loading';
