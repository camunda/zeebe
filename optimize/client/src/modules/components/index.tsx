/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Button} from './Button';
export {Tabs} from './Tabs';
export {Icon} from './Icon';
export {ActionItem} from './ActionItem';
export {Input} from './Input';
export {LabeledInput} from './LabeledInput';
export {Labeled} from './Labeled';
export {Tooltip} from './Tooltip';
export {Modal} from './Modal';
export {Dropdown} from './Dropdown';
export {Select} from './Select';
export {TextEditor} from './TextEditor';
export {PageTitle} from './PageTitle';
export {EntityDescription} from './EntityDescription';
export {EntityNameForm} from './EntityNameForm';
export {LoadingIndicator} from './LoadingIndicator';
export {DocsLink} from './DocsLink';
export {UserTypeahead} from './UserTypeahead';
export {BPMNDiagram, ClickBehavior} from './BPMNDiagram';
export {DMNDiagram} from './DMNDiagram';
export {DownloadButton} from './DownloadButton';
export {NoDataNotice} from './ReportRenderer';
export {Table} from './Table';
export {TargetValueBadge} from './TargetValueBadge';
export {DatePicker} from './DatePicker';
export {DateRangeInput} from './DateRangeInput';
export {Checklist} from './Checklist';
export {DurationChart} from './DurationChart';
export {ColorPicker} from './ColorPicker';
export {Popover} from './Popover';
export {TenantInfo} from './TenantInfo';
export {VersionPopover, TenantPopover} from './DefinitionSelection';
export {CopyToClipboard} from './CopyToClipboard';
export {ShareEntity} from './ShareEntity';
export {ErrorPage} from './ErrorPage';
export {ErrorBoundary} from './ErrorBoundary';
export {TargetSelection} from './TargetSelection';
// @ts-ignore
export {DefinitionSelection} from './DefinitionSelection';
// @ts-ignore
export {ReportRenderer} from './ReportRenderer';
export {Loading} from './Loading';
export {EntityList} from './EntityList';
export {EmptyState} from './EmptyState';

export type {InputProps} from './Input';
export type {LabeledProps} from './Labeled';
export type {User} from './UserTypeahead';
export type {Canvas, ModdleElement, RegistryElement} from './BPMNDiagram';
export type {TooltipProps} from './Tooltip';
export type {TableBody, TableHead} from './Table';
