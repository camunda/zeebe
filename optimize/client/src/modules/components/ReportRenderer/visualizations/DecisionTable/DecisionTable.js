/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React, {useState, useCallback, useMemo} from 'react';

import {DMNDiagram} from 'components';
import {themed} from 'theme';
import {formatters} from 'services';

import DmnJsPortal from './DmnJsPortal';
import createHitsColumnAddon from './HitsColumnAddon';

import './DecisionTable.scss';

export function DecisionTable({report, theme}) {
  const {
    result: {data, instanceCount},
    data: {
      definitions: [{key}],
      configuration: {xml, hideAbsoluteValue, hideRelativeValue, showGradientBars},
    },
  } = report;

  const resultObj = useMemo(() => formatters.objectifyResult(data), [data]);
  const [entryPoints, setEntryPoints] = useState({rules: {}});

  const renderRuleCell = (ruleId) => {
    const resultNumber = resultObj[ruleId] || 0;
    const percentage = Math.round((resultNumber / instanceCount) * 1000) / 10 || 0;

    const node = entryPoints.rules[ruleId];
    if (showGradientBars) {
      const progress = resultNumber / instanceCount;

      node.style.background = `linear-gradient(to right, ${getColor(theme, 0)} 0%, ${getColor(
        theme,
        progress
      )} ${percentage}%, transparent ${percentage}%)`;
    } else {
      node.style.background = '';
    }

    let outputString = `${resultNumber} (${percentage}%)`;
    if (hideAbsoluteValue && hideRelativeValue) {
      outputString = '';
    } else if (hideAbsoluteValue) {
      outputString = percentage + '%';
    } else if (hideRelativeValue) {
      outputString = resultNumber;
    }

    return (
      <DmnJsPortal key={ruleId} renderIn={entryPoints.rules[ruleId]}>
        <i>{outputString}</i>
      </DmnJsPortal>
    );
  };

  const {rules, summary} = entryPoints;
  const hitCount = data.map(({value}) => value).reduce((sum, current) => sum + current, 0);

  const hitsColumn = useMemo(createHitsColumnAddon, []);
  const additionalModules = useMemo(() => [hitsColumn.Addon], [hitsColumn.Addon]);
  const onLoad = useCallback(() => {
    setEntryPoints(hitsColumn.entryPoints);
  }, [hitsColumn.entryPoints]);

  return (
    <div className="DecisionTable">
      <DMNDiagram
        xml={xml}
        decisionDefinitionKey={key}
        additionalModules={additionalModules}
        onLoad={onLoad}
      >
        {Object.keys(rules).map(renderRuleCell)}
        <DmnJsPortal renderIn={summary}>
          <b>
            {instanceCount} Evaluation{instanceCount !== 1 ? 's' : ''}
            {hitCount > instanceCount && ` / ${hitCount} Hits`}
          </b>
        </DmnJsPortal>
      </DMNDiagram>
    </div>
  );
}

export default themed(DecisionTable);

function getColor(theme, progress) {
  if (theme === 'light') {
    return `hsl(223, 100%, ${90 - progress * 30}%)`;
  } else {
    return `hsl(223, 100%, ${30 + progress * 30}%)`;
  }
}
