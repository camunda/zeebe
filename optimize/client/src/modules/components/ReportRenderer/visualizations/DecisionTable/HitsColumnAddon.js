/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {Cell} from 'table-js/lib/components';

import {createComponentVNode, createVNode, createTextVNode, Component} from 'inferno';

export default function HitsColumnAddon() {
  const entryPoints = {rules: {}};

  const CustomCell = class CustomCell extends Component {
    componentDidMount() {
      if (this.props.summary) {
        entryPoints.summary = this.el;
      } else {
        entryPoints.rules[this.props.row.id] = this.el;
      }
    }

    render() {
      return createComponentVNode(2, Cell, {
        ref: (el) => (this.el = el),
        className: 'CustomCell',
      });
    }
  };

  function addHitsColumn(components) {
    components.onGetComponent('table.foot', () => (props) => {
      const emptyCells = [];
      for (let i = 0; i < props.cols.length + 1; i++) {
        emptyCells.push(createVNode(1, 'td'));
      }
      return createVNode(
        1,
        'tfoot',
        null,
        createVNode(
          1,
          'tr',
          null,
          [
            createVNode(1, 'td', null, createTextVNode('Total'), 2),
            ...emptyCells,
            createComponentVNode(2, CustomCell, {summary: true}),
          ],
          0
        ),
        2
      );
    });
    components.onGetComponent('cell', ({cellType}) => {
      if (cellType === 'after-label-cells') {
        return () => createVNode(1, 'th', 'hit header', createTextVNode('Hits'), 2);
      }
      if (cellType === 'after-rule-cells') {
        return CustomCell;
      }
    });
  }
  addHitsColumn.$inject = ['components'];

  return {
    entryPoints,
    Addon: {
      __init__: ['hitsColumn'],
      hitsColumn: ['type', addHitsColumn],
    },
  };
}
