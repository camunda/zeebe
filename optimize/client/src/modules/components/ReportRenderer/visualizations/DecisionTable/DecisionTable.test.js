/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import {DMNDiagram} from 'components';

import {DecisionTable} from './DecisionTable';
import createHitsColumnAddon from './HitsColumnAddon';

jest.mock('./HitsColumnAddon');

createHitsColumnAddon.mockReturnValue({
  entryPoints: {
    rules: {
      a: document.createElement('td'),
      b: document.createElement('td'),
    },
    summary: document.createElement('td'),
  },
  Addon: 'addon',
});

const props = {
  report: {
    data: {
      configuration: {
        xml: 'dmn xml string',
      },
      definitions: [
        {
          key: 'key',
        },
      ],
    },
    result: {
      instanceCount: 3,
      data: [
        {key: 'a', value: 1},
        {key: 'b', value: 2},
      ],
    },
  },
  mightFail: jest.fn().mockImplementation((data, cb) => cb(data)),
};

it('should render content in DmnJsPortals', () => {
  const node = shallow(<DecisionTable {...props} />);

  node.find(DMNDiagram).prop('onLoad')();

  expect(node).toMatchSnapshot();
});

it('should display meaningful data if there are no evaluations', () => {
  const node = shallow(
    <DecisionTable
      {...props}
      report={{data: props.report.data, result: {instanceCount: 0, data: []}}}
    />
  );

  node.find(DMNDiagram).prop('onLoad')();

  expect(node).toMatchSnapshot();
});
