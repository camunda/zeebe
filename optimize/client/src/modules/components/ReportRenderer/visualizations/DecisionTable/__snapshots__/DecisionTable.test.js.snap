// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display meaningful data if there are no evaluations 1`] = `
<div
  className="DecisionTable"
>
  <DMNDiagramErrorHandler
    additionalModules={
      Array [
        "addon",
      ]
    }
    decisionDefinitionKey="key"
    onLoad={[Function]}
    xml="dmn xml string"
  >
    <DmnJsPortal
      key="a"
      renderIn={<td />}
    >
      <i>
        0 (0%)
      </i>
    </DmnJsPortal>
    <DmnJsPortal
      key="b"
      renderIn={<td />}
    >
      <i>
        0 (0%)
      </i>
    </DmnJsPortal>
    <DmnJsPortal
      renderIn={<td />}
    >
      <b>
        0
         Evaluation
        s
      </b>
    </DmnJsPortal>
  </DMNDiagramErrorHandler>
</div>
`;

exports[`should render content in DmnJsPortals 1`] = `
<div
  className="DecisionTable"
>
  <DMNDiagramErrorHandler
    additionalModules={
      Array [
        "addon",
      ]
    }
    decisionDefinitionKey="key"
    onLoad={[Function]}
    xml="dmn xml string"
  >
    <DmnJsPortal
      key="a"
      renderIn={<td />}
    >
      <i>
        1 (33.3%)
      </i>
    </DmnJsPortal>
    <DmnJsPortal
      key="b"
      renderIn={<td />}
    >
      <i>
        2 (66.7%)
      </i>
    </DmnJsPortal>
    <DmnJsPortal
      renderIn={<td />}
    >
      <b>
        3
         Evaluation
        s
      </b>
    </DmnJsPortal>
  </DMNDiagramErrorHandler>
</div>
`;
